src/impl_/coroutine.rs

```diff
@@ -16,3 +24,67 @@ where
 {
     Coroutine::new(Some(name.into()), qualname_prefix, throw_callback, future)
 }
+
+fn get_ptr<T: PyClass>(obj: &Py<T>) -> *mut T {
+    // SAFETY: Py<T> can be casted as *const PyCell<T>
+    unsafe { &*(obj.as_ptr() as *const PyCell<T>) }.get_ptr()
+}
+
+pub struct RefGuard<T: PyClass>(Py<T>);
```

Comment:
To simplify the implementation, I wonder if this should be

```suggestion
pub struct RefGuard<T: PyClass>(ManuallyDrop<PyRef<'static, T>>);
```

Reply:
If you want to use `PyRef` directly, you should not use `ManuallyDrop`, because you need to release the borrow with `PyRef` destructor.
Also, `PyRef` doesn't own the object, so you have to also capture `Py<T>` in the future, to prevent the object dealloacation. And you have to be careful to the drop order, because if `Py<T>` is dropped before `PyRef`, the associated object could be deallocated before running `PyRef` destructor, and you could crash.
Last but not least, `PyRef` is not `Sent`, so you have to implement `Send` manually for `RefGuard`.

All in all, I'm not sure that's simpler, especially taking in account the crash potential with the destructor order.
I've just rewritten `RefGuard`/`RefMutGuard` constructors which was overly complex.

Reply:
All great points why my suggestion is no good 👍